[LoopInterchange] Make the entries of the Dependency Matrix unique

sjoerdmeijer opened the pull request:
The entries in the dependency matrix can contain a lot of duplicates, which is unnecessary and results in more checks that we can avoid, and this patch adds that.

I haven't added tests because the printing of the dependency matrix is guarded by an #ifdef DUMP_DEP_MATRICES so cannot be printed in normal builds. But all existing regression tests are passing, so this is tested in that way.

sjoerdmeijer committed:
[LoopInterchange] Make the entries of the Dependency Matrix unique

The entries in the dependency matrix can contain a lot of duplicates,
which is unnecessary and results in more checks that we can avoid, and
this patch does that.

madhur13490 commented on llvm/lib/Transforms/Scalar/LoopInterchange.cpp:
```diff
@@ -110,6 +111,7 @@ static bool populateDependencyMatrix(CharMatrix &DepMatrix, unsigned Level,
                     << " Loads and Stores to analyze\n");
 
   ValueVector::iterator I, IE, J, JE;
+  std::unordered_set<std::string> UniqueDepEntries;
```
Using `unordered_set` is frowned upon as per [manual ](https://llvm.org/docs/ProgrammersManual.html#other-set-like-container-options)

madhur13490 commented:
> I haven't added tests because the printing of the dependency matrix is guarded by an #ifdef DUMP_DEP_MATRICES so cannot be printed in normal builds. But all existing regression tests are passing, so this is tested in that way.

1) I think we should print the dependency matrix in debug mode too. #ifdef DUMP_DEP_MATRICES is unusual to me too. We can dump things in debug mode for better debugging ability.
2) I am curious. Since, now the code would be doing less checks, do we see any improvement in compile time?

sjoerdmeijer commented:
I have:
- removed the `fdef DUMP_DEP_MATRICIES`, replaced it with LLVM_DEBUG because it is indeed very useful to see this with debug enabled.
- replaced the unordered set with DenseSmallSet. The number of entries isn't very big normally, so I don't expect this to matter very much, but indeed DenseSmallSet is a bit better and also more consistent. 

In terms of overall compile-time, I don't expect this to make a big impact in the grand scheme of things. It may help a little, but might be difficult to measure. I see this more as a clean up. For one thing, just seeing the dependency matrix being dumped with -debug and not seeing all the duplicates is an improvement.

mshockwave commented on llvm/lib/Transforms/Scalar/LoopInterchange.cpp:
```diff
@@ -110,6 +110,7 @@ static bool populateDependencyMatrix(CharMatrix &DepMatrix, unsigned Level,
                     << " Loads and Stores to analyze\n");
 
   ValueVector::iterator I, IE, J, JE;
+  SmallDenseSet<size_t> Seen;
```
is there any reason we can't use `SmallDenseSet<StringRef>` here? Manually calculate hashes on common types like std::string / StringRef looks a little unusual to me.

sjoerdmeijer replied:
StringRef doesn't own the data. I construct a std::string in the loop from a vector of characters. My concern is that when that string goes out of scope, the std::string is deallocated, and the StringRef in the set points to invalid memory. I thought just storing an integer would be easiest. And I did a grep for "DenseSmallSet<std::string" and didn't see any precedent for that. Let me know if I got any of this wrong, am of course happy to change this.

nikic replied:
For a string set that owns the strings you would normally use StringSet.

sjoerdmeijer replied:
Ah, okay, thanks!

mshockwave commented on llvm/lib/Transforms/Scalar/LoopInterchange.cpp:
```diff
@@ -156,7 +157,14 @@ static bool populateDependencyMatrix(CharMatrix &DepMatrix, unsigned Level,
+        // Make sure we only add unique entries to the dependency matrix.
+        size_t Hash =
+            std::hash<std::string>{}(std::string(Dep.begin(), Dep.end()));
+        if (!Seen.count(Hash)) {
+          Seen.insert(Hash);
```
```suggestion
        if (Seen.insert(Hash).second) {
```
The second element of the returned std::pair from insert indicate if an insertion actually happens (i.e. it's first time you see it)

mshockwave commented:
Now the print function is guarded by NDEBUG instead, I believe you can add a test now. Please remember to add `REQUIRES: asserts` to the test.

sjoerdmeijer commented:
- Now using StringSet,
- And have added a test.

nikic commented on llvm/lib/Transforms/Scalar/LoopInterchange.cpp:
```diff
@@ -156,7 +158,13 @@ static bool populateDependencyMatrix(CharMatrix &DepMatrix, unsigned Level,
 
-        DepMatrix.push_back(Dep);
+        // Make sure we only add unique entries to the dependency matrix.
+        std::string DepStr = std::string(Dep.begin(), Dep.end());
```
I think you should be able to create this as a StringRef. No need to allocate std::string here.

nikic commented on llvm/lib/Transforms/Scalar/LoopInterchange.cpp:
```diff
@@ -156,7 +158,13 @@ static bool populateDependencyMatrix(CharMatrix &DepMatrix, unsigned Level,
+        // Make sure we only add unique entries to the dependency matrix.
+        std::string DepStr = std::string(Dep.begin(), Dep.end());
+        if (!Seen.count(DepStr)) {
+          Seen.insert(DepStr);
```
You should directly call insert and check the return value. That saves one hash table lookup.

sjoerdmeijer commented:
Thanks for the suggestion:
- now using `insert` to check if the insertion happened.

mshockwave reviewed (approved):
LGTM

nikic commented on llvm/lib/Transforms/Scalar/LoopInterchange.cpp:
```diff
@@ -156,7 +158,10 @@ static bool populateDependencyMatrix(CharMatrix &DepMatrix, unsigned Level,
 
-        DepMatrix.push_back(Dep);
+        // Make sure we only add unique entries to the dependency matrix.
+        if (Seen.insert(std::string(Dep.begin(), Dep.end())).second) 
```
```suggestion
        if (Seen.insert(StringRef(Dep.data(), Dep.size())).second) 
```
Doesn't something like this work?

sjoerdmeijer replied:
Thanks, will check and apply this before committing this, which I assume is okay.
